pkg/jobmanager/job.go

```diff
@@ -0,0 +1,258 @@
+	"github.com/adalton/teleport-exercise/pkg/io"
+	"github.com/google/uuid"
+)
+
+// JobStatus models the current status of a job.
+type JobStatus struct {
+	Owner     string
+	Name      string
+	Id        string
+	Running   bool
+	Pid       int
+	ExitCode  int
+	SignalNum syscall.Signal
+	RunError  error
+}
+
+// Job defines an interface for objects that behave like jobs.  This enables
+// us to define both a production job type as well a a job type for unit
+// testing.
+type Job interface {
+	Start() error
+	Stop() error
+	Status() *JobStatus
+	StdoutStream() *io.ByteStream
+	StderrStream() *io.ByteStream
+	Name() string
+	Id() uuid.UUID
+}
+
+// concreteJob implements the Job interface and provides the production implementation
+// of concreteJob behavior.
+type concreteJob struct {
+	mutex         sync.Mutex
+	owner         string
+	id            uuid.UUID
+	name          string
+	cgControllers []cgroup.Controller
+	programName   string
+	programArgs   []string
+	cmd           *exec.Cmd
+	stdoutBuffer  io.OutputBuffer
+	stderrBuffer  io.OutputBuffer
+	running       bool
+	runErrors     []error
+}
+
+// NewJob creates and returns a new concreteJob based on the given values.
+func NewJob(
+	owner string,
+	name string,
+	cgControllers []cgroup.Controller,
+	programName string,
+	programArgs ...string,
+) Job {
+
+	return NewJobDetailed(
+		owner,
+		name,
+		cgControllers,
+		io.NewMemoryBuffer(),
+		io.NewMemoryBuffer(),
+		programName,
+		programArgs...,
+	)
+}
+
+func NewJobDetailed(
+	owner string,
+	name string,
+	cgControllers []cgroup.Controller,
+	stdoutBuffer io.OutputBuffer,
+	stderrBuffer io.OutputBuffer,
+	programName string,
+	programArgs ...string,
+) Job {
+
+	return &concreteJob{
+		owner:         owner,
+		id:            uuid.New(),
+		name:          name,
+		cgControllers: cgControllers,
+		programName:   programName,
+		programArgs:   programArgs,
+		stdoutBuffer:  stdoutBuffer,
+		stderrBuffer:  stderrBuffer,
+	}
+}
+
+// Start starts the job if it hasn't already been started.
+func (j *concreteJob) Start() error {
+	j.mutex.Lock()
+	defer j.mutex.Unlock()
+
+	if j.cmd != nil {
+		return fmt.Errorf("job %s (%v) has already been started", j.name, j.id)
+	}
+
+	cgroupSet := cgroup.NewSet(j.id, j.cgControllers...)
+	if err := cgroupSet.Create(); err != nil {
+		return err
+	}
+
+	args := cgroupSet.TaskFiles()
+	args = append(args, "--")
+	args = append(args, j.programName)
+	args = append(args, j.programArgs...)
+
+	j.cmd = exec.Command(config.CgexecPath, args...)
+	j.cmd.Stdout = j.stdoutBuffer
+	j.cmd.Stderr = j.stderrBuffer
+	j.cmd.Env = make([]string, 0) // Do not pass along our environment
+
+	//j.cmd.Dir = "/" // If we were to chroot
+
+	j.cmd.SysProcAttr = &syscall.SysProcAttr{
+		Chroot: "", // This would be non-empty to actually do a chroot
+		Cloneflags: syscall.CLONE_NEWPID |
+			syscall.CLONE_NEWNS |
+			syscall.CLONE_NEWNET,
+	}
+
+	j.running = true
+
+	go func() {
+		defer j.lockedOperation(func() {
```

Comment:
In the happy path they are sequential.  My rationale for doing this was to guard against some panic causing us not to get to the end of the function -- I wanted to make sure that the cgroups were destroyed in that case.  Granted as the code stands now, there's little room for such a panic, so maybe this is overkill.

Reply:
Yeah, I don't see the point for grabbing the lock twice. I'd either defer everything, or more likely push it all to the end, since we the function is short and doesn't have multiple exits.

Reply:
Yeah, I don't see the point for grabbing the lock twice. I'd either defer everything, or more likely push it all to the end, since we the function is short and doesn't have multiple exits.